File: app/src/main/java/com/petnbu/petnbu/notification/NotificationsRecyclerViewAdapter.java
package com.petnbu.petnbu.notification;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.os.AsyncTask;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextUtils;
import android.text.format.DateUtils;
import android.text.style.ForegroundColorSpan;
import android.text.style.StyleSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.common.internal.Preconditions;
import com.petnbu.petnbu.BaseBindingViewHolder;
import com.petnbu.petnbu.GlideApp;
import com.petnbu.petnbu.GlideRequests;
import com.petnbu.petnbu.R;
import com.petnbu.petnbu.databinding.ViewNotificationBinding;
import com.petnbu.petnbu.feed.FeedsRecyclerViewAdapter;
import com.petnbu.petnbu.model.FeedUI;
import com.petnbu.petnbu.model.Notification;
import com.petnbu.petnbu.model.NotificationUI;
import com.petnbu.petnbu.util.ColorUtils;
import com.petnbu.petnbu.util.Objects;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class NotificationsRecyclerViewAdapter extends RecyclerView.Adapter<BaseBindingViewHolder> {

    private final int VIEW_TYPE_NOTIFICATION = 1;
    private final int VIEW_TYPE_LOADING = 2;

    private GlideRequests mGlideRequests;
    private List<NotificationUI> mNotifications;
    private boolean mAddLoadMore;
    private int mDataVersion = 0;

    public NotificationsRecyclerViewAdapter(Context context, List<NotificationUI> notifications) {
        Preconditions.checkNotNull(context);
        mGlideRequests = GlideApp.with(context);
        mNotifications = notifications != null ? notifications : new ArrayList<>(0);
    }

    @NonNull
    @Override
    public BaseBindingViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if(viewType == VIEW_TYPE_LOADING) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_loading, parent, false);
            return new ViewLoadingHolder(view);
        } else {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_notification, parent, false);
            return new ViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull BaseBindingViewHolder holder, int position) {
        holder.bindData(getItem(position));
    }

    @Override
    public int getItemCount() {
        return getDataItemCount() + (mAddLoadMore ? 1 : 0);
    }

    private int getDataItemCount() {
        return mNotifications.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (position < getDataItemCount() && getDataItemCount() > 0) {
            return VIEW_TYPE_NOTIFICATION;
        }
        return VIEW_TYPE_LOADING;
    }

    private NotificationUI getItem(int position) {
        if (position < 0 || position >= mNotifications.size())
            return null;
        return mNotifications.get(position);
    }

    public void setAddLoadMore(boolean addLoadMore) {
        if (mAddLoadMore != addLoadMore) {
            if (addLoadMore) {
                mAddLoadMore = addLoadMore;
                notifyItemInserted(getLoadingMoreItemPosition());
            } else {
                notifyItemRemoved(getLoadingMoreItemPosition());
                mAddLoadMore = addLoadMore;
            }
        }
    }

    private int getLoadingMoreItemPosition() {
        return mAddLoadMore ? getItemCount() - 1 : RecyclerView.NO_POSITION;
    }

    protected class ViewHolder extends BaseBindingViewHolder<ViewNotificationBinding, NotificationUI> {

        private NotificationUI mNotification;

        public ViewHolder(View itemView) {
            super(itemView);
        }

        @Override
        public void bindData(NotificationUI item) {
            mNotification = item;

            String avatarUrl = TextUtils.isEmpty(mNotification.getFromUser().getAvatar().getThumbnailUrl()) ?
                    mNotification.getFromUser().getAvatar().getOriginUrl() : mNotification.getFromUser().getAvatar().getThumbnailUrl();
            mGlideRequests
                    .load(avatarUrl)
                    .centerInside()
                    .into(mBinding.imgProfile);
            displayMessageContent();
            mBinding.tvDate.setText(DateUtils.getRelativeTimeSpanString(mNotification.getTimeCreated().getTime(),
                    Calendar.getInstance().getTimeInMillis(), 0L, DateUtils.FORMAT_ABBREV_RELATIVE));
        }

        @Override
        public void bindData(NotificationUI item, List<Object> payloads) {

        }

        private void displayMessageContent() {
            SpannableStringBuilder messageSpanBuilder = new SpannableStringBuilder(mNotification.getFromUser().getName());
            messageSpanBuilder.setSpan(new StyleSpan(Typeface.BOLD), 0, messageSpanBuilder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            messageSpanBuilder.append("  ");

            String message;
            Context context = itemView.getContext();
            switch (mNotification.getType()) {
                case Notification.TYPE_LIKE_FEED :
                    message = context.getString(R.string.notification_message_like, "post");
                    break;
                case Notification.TYPE_LIKE_COMMENT :
                    message = context.getString(R.string.notification_message_like, "comment");
                    break;
                case Notification.TYPE_LIKE_REPLY :
                    message =  context.getString(R.string.notification_message_like, "reply");
                    break;
                case Notification.TYPE_NEW_COMMENT :
                    message = context.getString(R.string.notification_message_comment);
                    break;
                case Notification.TYPE_NEW_REPLY :
                    message = context.getString(R.string.notification_message_reply);
                    break;
                default: message = "";
            }
            int start = messageSpanBuilder.length();
            messageSpanBuilder.append(message);
            messageSpanBuilder.setSpan(new ForegroundColorSpan(ColorUtils.modifyAlpha(Color.BLACK, 0.8f)), start, messageSpanBuilder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            mBinding.tvNotificationMessage.setText(messageSpanBuilder);
        }
    }

    protected class ViewLoadingHolder extends BaseBindingViewHolder<ViewNotificationBinding, NotificationUI> {

        public ViewLoadingHolder(View itemView) {
            super(itemView);
        }

        @Override
        public void bindData(NotificationUI item) {

        }

        @Override
        public void bindData(NotificationUI item, List<Object> payloads) {

        }
    }

    public static class NotificationDiffCallback extends DiffUtil.Callback{
        private final List<NotificationUI> oldData;
        private final List<NotificationUI> newData;

        public NotificationDiffCallback(List<NotificationUI> oldData, List<NotificationUI> newData) {
            this.oldData = oldData;
            this.newData = newData;
        }

        @Override
        public int getOldListSize() {
            return oldData.size();
        }

        @Override
        public int getNewListSize() {
            return newData.size();
        }

        @Override
        public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
            return oldData.get(oldItemPosition).getId().equals(newData.get(newItemPosition).getId());
        }

        @Override
        public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
            return Objects.equals(oldData.get(oldItemPosition), newData.get(newItemPosition));
        }
    }

    @SuppressLint("StaticFieldLeak")
    public void setData(List<NotificationUI> notifications) {
        mDataVersion++;
        final int startVersion = mDataVersion;
        new AsyncTask<Void, Void, DiffUtil.DiffResult>() {
            @Override
            protected DiffUtil.DiffResult doInBackground(Void... voids) {
                return DiffUtil.calculateDiff(new NotificationDiffCallback(mNotifications, notifications));
            }

            @Override
            protected void onPostExecute(DiffUtil.DiffResult diffResult) {
                if (startVersion != mDataVersion) {
                    // ignore update
                    return;
                }
                mNotifications = notifications;
                diffResult.dispatchUpdatesTo(NotificationsRecyclerViewAdapter.this);
            }
        }.execute();
    }
}

File: app/src/main/java/com/petnbu/petnbu/notification/NotificationsRecyclerViewAdapter.kt
package com.petnbu.petnbu.notification

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.support.v7.recyclerview.extensions.ListAdapter
import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.format.DateUtils
import android.text.style.ForegroundColorSpan
import android.text.style.StyleSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.petnbu.petnbu.BaseBindingViewHolder
import com.petnbu.petnbu.GlideApp
import com.petnbu.petnbu.GlideRequests
import com.petnbu.petnbu.R
import com.petnbu.petnbu.databinding.ViewNotificationBinding
import com.petnbu.petnbu.model.Notification
import com.petnbu.petnbu.model.NotificationUI
import com.petnbu.petnbu.util.ColorUtils
import com.petnbu.petnbu.util.Objects
import java.util.*

class NotificationsRecyclerViewAdapter(context: Context)
    : ListAdapter<NotificationUI, BaseBindingViewHolder<*, *>>(NotificationDiffCallback()) {

    private val glideRequests: GlideRequests = GlideApp.with(context)
    private var addLoadMore: Boolean = false

    private val loadingMoreItemPosition: Int
        get() = if (addLoadMore) itemCount - 1 else RecyclerView.NO_POSITION

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseBindingViewHolder<*, *> {
        return if (viewType == VIEW_TYPE_LOADING) {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.view_loading, parent, false)
            ViewLoadingHolder(view)
        } else {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.view_notification, parent, false)
            ViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: BaseBindingViewHolder<*, *>, position: Int) {
        val item: NotificationUI? = getItem(position)
        item?.run {
            (holder as? ViewHolder)?.bindData(this)
        }
    }

    override fun getItemCount() = super.getItemCount() + if (addLoadMore) 1 else 0

    override fun getItemViewType(position: Int) = if (position < itemCount && itemCount > 0) VIEW_TYPE_NOTIFICATION else VIEW_TYPE_LOADING

    fun setAddLoadMore(addLoadMore: Boolean) {
        if (this.addLoadMore != addLoadMore) {
            if (addLoadMore) {
                this.addLoadMore = addLoadMore
                notifyItemInserted(loadingMoreItemPosition)
            } else {
                notifyItemRemoved(loadingMoreItemPosition)
                this.addLoadMore = addLoadMore
            }
        }
    }

    private inner class ViewHolder(itemView: View) : BaseBindingViewHolder<ViewNotificationBinding, NotificationUI>(itemView) {

        private lateinit var notification: NotificationUI

        override fun bindData(item: NotificationUI) {
            notification = item

            val avatarUrl = if (!notification.fromUser.avatar.thumbnailUrl.isNullOrEmpty())
                notification.fromUser.avatar.thumbnailUrl
            else
                notification.fromUser.avatar.originUrl
            glideRequests
                    .load(avatarUrl)
                    .centerInside()
                    .into(mBinding.imgProfile)
            displayMessageContent()
            mBinding.tvDate.text = DateUtils.getRelativeTimeSpanString(notification.timeCreated.time,
                    Calendar.getInstance().timeInMillis, 0L, DateUtils.FORMAT_ABBREV_RELATIVE)
        }

        override fun bindData(item: NotificationUI, payloads: List<Any>) {

        }

        private fun displayMessageContent() {
            val context = itemView.context
            val messageSpanBuilder = SpannableStringBuilder(notification.fromUser.name)
            val message = when (notification.type) {
                Notification.TYPE_LIKE_FEED -> context.getString(R.string.notification_message_like, "post")
                Notification.TYPE_LIKE_COMMENT -> context.getString(R.string.notification_message_like, "comment")
                Notification.TYPE_LIKE_REPLY -> context.getString(R.string.notification_message_like, "reply")
                Notification.TYPE_NEW_COMMENT -> context.getString(R.string.notification_message_comment)
                Notification.TYPE_NEW_REPLY -> context.getString(R.string.notification_message_reply)
                else -> ""
            }

            messageSpanBuilder.apply {
                setSpan(StyleSpan(Typeface.BOLD), 0, messageSpanBuilder.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                append("  ")
                val start = length
                append(message)
                setSpan(ForegroundColorSpan(ColorUtils.modifyAlpha(Color.BLACK, 0.8f)), start, messageSpanBuilder.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            mBinding.tvNotificationMessage.text = messageSpanBuilder
        }
    }

    private class ViewLoadingHolder(itemView: View) : BaseBindingViewHolder<ViewNotificationBinding, NotificationUI>(itemView) {

        override fun bindData(item: NotificationUI) {

        }

        override fun bindData(item: NotificationUI, payloads: List<Any>) {

        }
    }

    private class NotificationDiffCallback : DiffUtil.ItemCallback<NotificationUI>() {

        override fun areContentsTheSame(oldItem: NotificationUI, newItem: NotificationUI) = oldItem.id == newItem.id

        override fun areItemsTheSame(oldItem: NotificationUI, newItem: NotificationUI) = Objects.equals(oldItem, newItem)
    }

    companion object {
        private const val VIEW_TYPE_NOTIFICATION = 1
        private const val VIEW_TYPE_LOADING = 2
    }
}
